ilp.py

```diff
@@ -0,0 +1,187 @@
+import pulp
+
+
+def find_fence_placement(edges, cycles, node_weights, edge_weights):
+    """
+    Function to find the optimal fence placement for given critical cycles using ILP.
+
+    Parameters:
+    critical_cycles (list of list of tuples): List of critical cycles, where each cycle is a list of edges (tuples).
+    node_weights (dict): Dictionary mapping node indices to their weights, indicating if they are Read or Write operations.
+
+    Returns:
+    dict: Dictionary with edges as keys and binary values indicating whether a fence is placed (1) or not (0).
+    """
+
+    filtered_edges = find_relevant_edges(edges, cycles, node_weights, edge_weights)
+
+    # Extract unique edges from all critical cycles
+    unique_edges = set(filtered_edges)
+
+    # Initialize the ILP problem
+    prob = pulp.LpProblem("TSOFencePlacement", pulp.LpMinimize)
+
+    # Define the decision variables for each unique edge
+    fences = {edge: pulp.LpVariable(f'f_{edge}', cat='Binary') for edge in unique_edges}
+
+    # Objective function: Minimize the number of fences
+    prob += pulp.lpSum(fences[edge] for edge in fences), "MinimizeFences"
+
+    # Constraints: Ensure each critical cycle is broken
+    for i, cycle in enumerate(cycles):
+        cycle_edges = []
+        for i in range(len(cycle)):
+            for j in range(len(cycle)):
+                if i != j:
+                    u = cycle[i]
+                    v = cycle[j]
+                    cycle_edges.append((u, v))
+        prob += pulp.lpSum(fences[edge] for edge in cycle_edges if edge in fences) >= 1, f"BreakCriticalCycle_{i}"
+    # Solve the problem
+    prob.solve()
+
+    # Prepare the results
+    result = {edge: int(pulp.value(fences[edge])) for edge in fences}
+
+    return result
+
+
+def find_relevant_edges(edges, cycles, node_weights, edge_weights):
```

Comment:
It might be nice to do this in rust as a form of preprocessing, but we can see if it becomes a bottleneck